File: app/src/main/java/com/chichiangho/base/base/BaseTabActivity.java
package com.chichiangho.base.base;

import android.os.Bundle;
import android.support.annotation.ColorRes;
import android.support.annotation.DimenRes;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.support.v4.view.ViewPager.OnPageChangeListener;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.chichiangho.base.R;
import com.chichiangho.base.utils.DeviceUtil;
import com.chichiangho.base.widgets.TabViewPager;

import java.util.ArrayList;

public class BaseTabActivity extends BaseTitleActivity {

    private MyAdapter mAdapter;
    private TabViewPager mViewPager;
    private LinearLayout mTab;

    private ArrayList<Holder> mList = new ArrayList<>(4);
    private int mTabBgColor = R.color.white;
    private int mTabHeight = R.dimen.tab_height;
    private int mTextSize = R.dimen.size_middle;
    private int mTextNormal = R.color.tab_text_normal;
    private int mTextChecked = R.color.tab_text_checked;
    private Holder cur;
    private boolean needDivider;
    private boolean mNeedUnderHint;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_tab);

        mViewPager = (TabViewPager) findViewById(R.id.view_pager);
        mAdapter = new MyAdapter(getSupportFragmentManager());
        mViewPager.setOffscreenPageLimit(3);
        mViewPager.setAdapter(mAdapter);
        mViewPager.addOnPageChangeListener(new OnPageChangeListener() {

            @Override
            public void onPageSelected(int arg0) {
                onViewPagerSelected(arg0);
            }

            @Override
            public void onPageScrolled(int arg0, float arg1, int arg2) {
            }

            @Override
            public void onPageScrollStateChanged(int arg0) {
            }
        });
    }

    protected void onViewPagerSelected(int arg0) {
        if (cur != null) {
            cur.getTv().setTextColor(getResources().getColor(mTextNormal));
            cur.getIv().setImageResource(cur.getIn());
            if (mNeedUnderHint)
                cur.getUnder().setVisibility(View.GONE);
        }
        cur = mList.get(arg0);
        cur.getTv().setTextColor(getResources().getColor(mTextChecked));
        cur.getIv().setImageResource(cur.getIs());
        if (mNeedUnderHint)
            cur.getUnder().setVisibility(View.VISIBLE);
    }

    public BaseTabActivity prepare(@DimenRes int tabHeight, @DimenRes int textSize, @ColorRes int textNormalColor, @ColorRes int textCheckedColor, @ColorRes int bgColor, boolean needUnderHint) {
        mTabHeight = tabHeight;
        mTextSize = textSize;
        mTextNormal = textNormalColor;
        mTextChecked = textCheckedColor;
        mTabBgColor = bgColor;
        mNeedUnderHint = needUnderHint;
        return this;
    }

    public BaseTabActivity addTab(Fragment fragment, int tabTextRes) {
        return addTab(fragment, tabTextRes, 0, 0);
    }

    public BaseTabActivity addTab(Fragment fragment, int tabTextRes, int imgNormal, int imgSelected) {
        Holder holder = new Holder(LayoutInflater.from(this).inflate(R.layout.layout_tab_item, null), fragment, imgNormal, imgSelected);
        mList.add(holder);
        holder.getTv().setText(tabTextRes);
        holder.getV().setOnClickListener(new OnClickListener() {
            @Override
            public void onClick(View v) {
                for (int i = 0; i < mList.size(); i++) {
                    if (v == mList.get(i).getV()) {
                        mViewPager.setCurrentItem(i);
                    }
                }
            }
        });
        return this;
    }

    public BaseTabActivity setNeedDivider(boolean needDivider) {
        this.needDivider = needDivider;
        return this;
    }

    public void commit() {
        if (mList.get(0).getIv().getVisibility() == View.GONE) {//if no img,which means tab at top
            mTab = (LinearLayout) findViewById(R.id.top_layout);
            findViewById(R.id.top_divider).setVisibility(View.VISIBLE);
        } else {
            mTab = (LinearLayout) findViewById(R.id.bottom_layout);
            findViewById(R.id.bottom_divider).setVisibility(View.VISIBLE);
        }
        mTab.setBackgroundColor(getResources().getColor(mTabBgColor));
        if (mList.size() > 1)
            mTab.setVisibility(View.VISIBLE);

        for (int i = 0; i < mList.size(); i++) {
            View tabItem = mList.get(i).getV();

            RelativeLayout.LayoutParams radioParams = new RelativeLayout.LayoutParams(DeviceUtil.getScreenWidth()
                    / mList.size(), (int) getResources().getDimension(mTabHeight));
            tabItem.setLayoutParams(radioParams);

            mTab.addView(tabItem);

            if (needDivider) {
                ImageView divider = new ImageView(getBaseContext());
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams((int) getResources().getDimension(R.dimen.divider_width), (int) (getResources().getDimension(mTabHeight) * 2 / 3));
                params.gravity = Gravity.CENTER;
                divider.setLayoutParams(params);
                divider.setBackgroundResource(R.color.divider_tab);
                mTab.addView(divider);
            }
            if (mNeedUnderHint) {
                RelativeLayout.LayoutParams params = new RelativeLayout.LayoutParams(
                        (DeviceUtil.getScreenWidth() / mList.size()) / 2, (int) getResources().getDimension(R.dimen.tab_divider_height));
                params.addRule(RelativeLayout.CENTER_HORIZONTAL);
                params.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM);
                mList.get(i).getUnder().setLayoutParams(params);
                mList.get(i).getUnder().setBackgroundResource(mTextChecked);
            } else {
                mList.get(i).getUnder().setVisibility(View.GONE);
            }
        }

        mAdapter.notifyDataSetChanged();
        onViewPagerSelected(0);
    }

    public TabViewPager getViewPager() {
        return mViewPager;
    }

    private class MyAdapter extends FragmentPagerAdapter {

        public MyAdapter(FragmentManager fm) {
            super(fm);
        }

        @Override
        public Fragment getItem(int arg0) {
            return mList.get(arg0).getF();
        }

        @Override
        public int getCount() {
            return mList.size();
        }

        @Override
        public void destroyItem(ViewGroup container, int position, Object object) {
        }
    }

    private class Holder {
        private Fragment f;
        private View v;
        private TextView tv;
        private ImageView iv;
        private View under;
        private int in;
        private int is;

        Holder(View v, Fragment f, int in, int is) {
            this.f = f;
            this.v = v;
            tv = (TextView) v.findViewById(R.id.text);
            tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, getResources().getDimensionPixelSize(mTextSize));
            iv = (ImageView) v.findViewById(R.id.image);
            under = v.findViewById(R.id.under_hint);
            this.in = in;
            this.is = is;

            if (in == 0 || is == 0)
                iv.setVisibility(View.GONE);
            else if (in != 0)
                iv.setImageResource(in);
        }


        public TextView getTv() {
            return tv;
        }

        public ImageView getIv() {
            return iv;
        }

        public int getIn() {
            return in;
        }

        public int getIs() {
            return is;
        }

        public Fragment getF() {
            return f;
        }

        public View getV() {
            return v;
        }

        public View getUnder() {
            return under;
        }

        public void setUnder(View under) {
            this.under = under;
        }
    }
}

File: common/src/main/java/com/chichiangho/common/base/BaseTabActivity.kt
package com.chichiangho.common.base

import android.os.Bundle
import android.support.annotation.ColorRes
import android.support.annotation.DimenRes
import android.support.v4.app.Fragment
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.ViewPager.OnPageChangeListener
import android.util.TypedValue
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import com.changhong.common.R
import com.chichiangho.common.extentions.parseColor
import kotlinx.android.synthetic.main.activity_tab.*
import java.util.*

abstract class BaseTabActivity : BaseTitleActivity() {

    private lateinit var mAdapter: MyAdapter
    private lateinit var mTab: LinearLayout

    private val mList = ArrayList<Holder>(4)
    private var mTabBgColor = R.color.white
    private var mTabHeight = R.dimen.tab_height
    private var mTextSize = R.dimen.size_middle
    private var mTextNormal = R.color.tab_text_normal
    private var mTextChecked = R.color.tab_text_checked
    private lateinit var cur: Holder
    private var needDivider: Boolean = false
    private var mNeedUnderHint: Boolean = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_tab)

        mAdapter = MyAdapter(supportFragmentManager)
        view_pager.offscreenPageLimit = 3
        view_pager.adapter = mAdapter
        view_pager.addOnPageChangeListener(object : OnPageChangeListener {

            override fun onPageSelected(arg0: Int) {
                onViewPagerSelected(arg0)
            }

            override fun onPageScrolled(arg0: Int, arg1: Float, arg2: Int) {}

            override fun onPageScrollStateChanged(arg0: Int) {}
        })
    }

    protected open fun onViewPagerSelected(arg0: Int) {
        cur.tv.setTextColor(parseColor(mTextNormal))
        cur.iv.setImageResource(cur.imgNormal)
        if (mNeedUnderHint)
            cur.under.visibility = View.GONE

        cur = mList[arg0]
        cur.tv.setTextColor(resources.getColor(mTextChecked))
        cur.iv.setImageResource(cur.imgSelected)
        if (mNeedUnderHint)
            cur.under.visibility = View.VISIBLE
    }

    fun prepare(@DimenRes tabHeight: Int, @DimenRes textSize: Int, @ColorRes textNormalColor: Int, @ColorRes textCheckedColor: Int, @ColorRes bgColor: Int, needUnderHint: Boolean): BaseTabActivity {
        mTabHeight = tabHeight
        mTextSize = textSize
        mTextNormal = textNormalColor
        mTextChecked = textCheckedColor
        mTabBgColor = bgColor
        mNeedUnderHint = needUnderHint
        return this
    }

    @JvmOverloads
    fun addTab(fragment: Fragment, tabTextRes: Int, imgNormal: Int = 0, imgSelected: Int = 0): BaseTabActivity {
        val holder = Holder(LayoutInflater.from(this).inflate(R.layout.layout_tab_item, null), fragment, imgNormal, imgSelected)
        mList.add(holder)
        holder.tv.setText(tabTextRes)
        holder.view.setOnClickListener { v ->
            mList.indices.filter { v === mList[it].view }.forEach { view_pager.currentItem = it }
        }
        return this
    }

    fun setNeedDivider(needDivider: Boolean): BaseTabActivity {
        this.needDivider = needDivider
        return this
    }

    fun commit() {
        if (mList[0].iv.visibility == View.GONE) {//if no img,which means tab at top
            mTab = top_layout
            top_divider.visibility = View.VISIBLE
        } else {
            mTab = bottom_layout
            bottom_divider.visibility = View.VISIBLE
        }
        mTab.setBackgroundColor(parseColor(mTabBgColor))
        if (mList.size > 1)
            mTab.visibility = View.VISIBLE

        for (i in mList.indices) {
            val tabItem = mList[i].view

            val radioParams = RelativeLayout.LayoutParams(Device.screenWidth / mList.size, resources.getDimension(mTabHeight).toInt())
            tabItem.layoutParams = radioParams

            mTab.addView(tabItem)

            if (needDivider) {
                val divider = ImageView(baseContext)
                val params = LinearLayout.LayoutParams(resources.getDimension(R.dimen.divider_width).toInt(), (resources.getDimension(mTabHeight) * 2 / 3).toInt())
                params.gravity = Gravity.CENTER
                divider.layoutParams = params
                divider.setBackgroundResource(R.color.divider_tab)
                mTab.addView(divider)
            }
            if (mNeedUnderHint) {
                val params = RelativeLayout.LayoutParams(Device.screenWidth / mList.size / 2, resources.getDimension(R.dimen.tab_divider_height).toInt())
                params.addRule(RelativeLayout.CENTER_HORIZONTAL)
                params.addRule(RelativeLayout.ALIGN_PARENT_BOTTOM)
                mList[i].under.layoutParams = params
                mList[i].under.setBackgroundResource(mTextChecked)
            } else {
                mList[i].under.visibility = View.GONE
            }
        }

        mAdapter.notifyDataSetChanged()
        cur = mList[0]
        onViewPagerSelected(0)
    }

    private inner class MyAdapter(fm: FragmentManager) : FragmentPagerAdapter(fm) {

        override fun getItem(arg0: Int): Fragment = mList[arg0].fragment

        override fun getCount(): Int = mList.size
    }

    private inner class Holder internal constructor(val view: View, val fragment: Fragment, val imgNormal: Int, val imgSelected: Int) {
        val tv: TextView = view.findViewById(R.id.text)
        val iv: ImageView = view.findViewById(R.id.image)
        var under: View = view.findViewById(R.id.under_hint)

        init {
            tv.setTextSize(TypedValue.COMPLEX_UNIT_PX, resources.getDimensionPixelSize(mTextSize).toFloat())

            if (imgNormal == 0 || imgSelected == 0)
                iv.visibility = View.GONE
            else if (imgNormal != 0)
                iv.setImageResource(imgNormal)
        }
    }
}
